docs/source/quickstart.rst

```diff
@@ -0,0 +1,422 @@
+           data = await stream.receive_some(1000)
+           if not data:
+               break
+           await stream.send_all(data)
+
+   # Usage: await trio.serve_tcp(echo_server_handler, ...)
+
+Now we need to test it, to make sure it's working correctly. In fact,
+since this is such complicated and sophisticated code, we're going to
+write lots of tests for it. And they'll all follow the same basic
+pattern: we'll start the echo server running in a background task,
+then connect to it, send it some test data, and see how it responds.
+Here's a first attempt::
+
+   # Let's cross our fingers and hope no-one else is using this port...
+   PORT = 14923
+
+   # Don't copy this -- we can do better
+   async def test_attempt_1():
+       async with trio.open_nursery() as nursery:
+           # Start server running in the background
+           nursery.start_soon(
+               partial(trio.serve_tcp, echo_server_handler, port=PORT)
+           )
+
+           # Connect to the server.
+           echo_client = await trio.open_tcp_stream("127.0.0.1", PORT)
+           # Send some test data, and check that it gets echoed back
+           async with echo_client:
+               for test_byte in [b"a", b"b", b"c"]:
+                   await echo_client.send_all(test_byte)
+                   assert await echo_client.receive_some(1) == test_byte
+
+This will mostly work, but it has a few problems. The most obvious one
+is that when we run it, even if everything works perfectly, it will
+hang at the end of the test – we never shut down the server, so the
+nursery block will wait forever for it to exit.
+
+To avoid this, we should cancel the nursery at the end of the test:
+
+.. code-block:: python3
+   :emphasize-lines: 7,20,21
+
+   # Let's cross our fingers and hope no-one else is using this port...
+   PORT = 14923
+
+   # Don't copy this -- we can do better
+   async def test_attempt_2():
+       async with trio.open_nursery() as nursery:
+           try:
+               # Start server running in the background
+               nursery.start_soon(
+                   partial(trio.serve_tcp, echo_server_handler, port=PORT)
+               )
+
+               # Connect to the server.
+               echo_client = await trio.open_tcp_stream("127.0.0.1", PORT)
+               # Send some test data, and check that it gets echoed back
+               async with echo_client:
+                   for test_byte in [b"a", b"b", b"c"]:
+                       await echo_client.send_all(test_byte)
+                       assert await echo_client.receive_some(1) == test_byte
+           finally:
+               nursery.cancel_scope.cancel()
+
+In fact, this pattern is *so* common, that pytest-trio provides a
+handy :data:`nursery` fixture to let you skip the boilerplate. Just
+add ``nursery`` to your test function arguments, and pytest-trio will
+open a nursery, pass it in to your function, and then cancel it for
+you afterwards:
+
+.. code-block:: python3
```

Comment:
This and other codeblocks aren't showing up in github's rst preview.  Doc issue or github issue??